src/components/Modal/GlobalModal.tsx

```diff
@@ -0,0 +1,86 @@
+import clsx from 'clsx';
+import type { PropsWithChildren } from 'react';
+import React, { useEffect, useRef } from 'react';
+import { FocusScope } from '@react-aria/focus';
+
+import { CloseIconRound } from './icons';
+
+import { useTranslationContext } from '../../context';
+import {
+  DialogPortalEntry,
+  modalDialogId,
+  useModalDialog,
+  useModalDialogIsOpen,
+} from '../Dialog';
+import type { ModalProps } from './Modal';
+
+export const GlobalModal = ({
+  children,
+  className,
+  onClose,
+  open,
+}: PropsWithChildren<ModalProps>) => {
+  const { t } = useTranslationContext('Modal');
+
+  const dialog = useModalDialog();
+  const isOpen = useModalDialogIsOpen();
+  const innerRef = useRef<HTMLDivElement | null>(null);
+  const closeRef = useRef<HTMLButtonElement | null>(null);
+
+  const handleClick = (event: React.MouseEvent<HTMLButtonElement | HTMLDivElement>) => {
+    if (innerRef.current?.contains(event.target as HTMLButtonElement | HTMLDivElement))
+      return;
+    onClose?.(event);
+    dialog.close();
+  };
+
+  useEffect(() => {
+    if (!isOpen) return;
+
+    const handleKeyDown = (event: KeyboardEvent) => {
+      if (event.key === 'Escape') {
+        onClose?.(event as unknown as React.KeyboardEvent);
+        dialog.close();
+      }
+    };
+
+    document.addEventListener('keydown', handleKeyDown);
+    return () => document.removeEventListener('keydown', handleKeyDown);
+  }, [dialog, onClose, isOpen]);
+
+  useEffect(() => {
+    if (open && !dialog.isOpen) {
+      dialog.open();
+    }
+  }, [dialog, open, children]);
```

Comment:
Are `children` required in the dependency array? This would make this effect run on every re-render.

Reply:
fixed: https://github.com/GetStream/stream-chat-react/pull/2792/commits/371670fb59ecb6b9aa87d04fb8bfef083e221527